internal/exp/metrics/staleness/staleness.go

```diff
@@ -0,0 +1,55 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package staleness
+
+import (
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/identity"
+)
+
+// We override how Now() is returned, so we can have deterministic tests
+var nowFunc = time.Now
+
+type Map[T any] interface {
+	Load(key identity.Stream) (T, bool)
+	Store(key identity.Stream, value T)
+	Delete(key identity.Stream)
+	// Range calls f sequentially for each key and value present in the map.
+	// If f returns false, range stops the iteration.
+	Range(f func(key identity.Stream, value T) bool)
```

Comment:
I think here is an opportunity to already conform to the future rangefunc api: https://go.dev/wiki/RangefuncExperiment

```go
func (m RawMap[T]) Items() iter.Seq2[identity.Stream, T] {
    return func(yield func(identity.Stream, T) bool) bool {
        for k, v := range m {
            if !yield(k, v) { break }
        }
        return false
    }
}
```

Reply:
That said, it looks like for `iter.Seq2` to be defined, we need go 1.22 and `GOEXPERIMENT=rangefunc`. Which I don't think we can allow for all the contrib.

Should we implement that ourselves for now? Or stick with Range until iter.Seq2 is GA?

Reply:
iter.Seq2 is merely an alias for `func(yield func(K, V) bool) bool`, so we can also just this raw type in the signature, which will be forward compatible!

Reply:
Ah good point. Thanks!